### PR TITLE
Rename `ItemKind::Ty` to `ItemKind::TyAlias`

### PR DESCRIPTION
**Pull request by varkor**
The current name is not entirely clear without context and `TyAlias` is consistent with `ItemKind::TraitAlias`.

### CODE REVIEW

File: src/librustc/hir/check_attr.rs [hunk 1]
```diff
@@ -75,7 +75,7 @@ impl Target {
             hir::ItemKind::Mod(..) => Target::Mod,
             hir::ItemKind::ForeignMod(..) => Target::ForeignMod,
             hir::ItemKind::GlobalAsm(..) => Target::GlobalAsm,
-            hir::ItemKind::Ty(..) => Target::Ty,
+            hir::ItemKind::TyAlias(..) => Target::Ty,
```
Comment (Centril) [inline src/librustc/hir/check_attr.rs]:
```suggestion
            hir::ItemKind::TyAlias(..) => Target::TyAlias,
```

File: src/librustc/hir/intravisit.rs [hunk 2]
```diff
@@ -500,7 +500,7 @@ pub fn walk_item<'v, V: Visitor<'v>>(visitor: &mut V, item: &'v Item) {
         ItemKind::GlobalAsm(_) => {
             visitor.visit_id(item.hir_id);
         }
-        ItemKind::Ty(ref ty, ref generics) => {
+        ItemKind::TyAlias(ref ty, ref generics) => {
```
Comment (Centril) [inline src/librustc/hir/intravisit.rs]: Better yet just `::TypeAlias`?

**Review by Centril — COMMENTED** [review]
I think we should rename `ImplItemKind::Type` to `::Ty(pe)Alias` also in that case.

File: src/librustc/hir/intravisit.rs (hunk 2 above)
Comment (varkor) [reply src/librustc/hir/intravisit.rs]: I think `Ty` is slightly more common in the codebase. We should make it consistent, though.

**Comment by Centril** [comment]
Strongly agree with this change; `::Ty` was pretty cryptic!

r=me when-green

**Comment by eddyb** [comment]
> I think we should rename ImplItemKind::Type to ::Ty(pe)Alias also in that case.

No, that's an associated type...

File: src/librustc/hir/mod.rs [hunk 3]
```diff
@@ -1837,7 +1837,7 @@ pub enum ImplItemKind {
     /// A method implementation with the given signature and body.
     Method(MethodSig, BodyId),
     /// An associated type.
-    Type(P<Ty>),
+    TyAlias(P<Ty>),
```
Comment (eddyb) [inline src/librustc/hir/mod.rs]: Basically I think this should be reverted. Also, as per some issue, `Method` should be `Fn`.

**Comment by Centril** [comment]
> No, that's an associated type...

The full name is "associated type alias/synonym".
"Associated type" is a shorthand.

See [“Type Associated Type Synonyms”. M. Chakravarty, G. Keller, and S. Peyton Jones. In Proceedings of “The Tenth ACM SIGPLAN International Conference on Functional Programming”, ACM Press, pages 241-253, 2005](https://www.microsoft.com/en-us/research/wp-content/uploads/2005/01/at-syns.pdf).

**Comment by eddyb** [comment]
@Centril I've never heard "associated type alias" in the context of Rust, and also this PR only changes the `impl`-associated item, but not the `trait`-associated one.

**Comment by Centril** [comment]
Ah woops; we should change the `trait` one as well then since consistency is key.

As for "in the context of Rust", that's probably because it is just shorter to use the abbreviation. Folks in the Haskell / C++ communities usually don't say the full thing either.

Another suggestion for you if you don't like `::TyAlias`: Let's consistently use `::Type` for `ItemKind`, `ImplItemKind`, `TraitItemKind`.

**Comment by eddyb** [comment]
I like `Type` because it matches the surface syntax (same with `Const`/`Static`/`Fn`).

**Comment by Centril** [comment]
Let's also rename `OpaqueTy` to `OpaqueType` in that case.

@eddyb I'll happily review that PR :)

**Comment by varkor** [comment]
I prefer `TyAlias` (or similar) over `Ty` (though `AssocTy` is also fine), because `Ty` is so overloaded and to someone unfamiliar with a particular enum, it can be unclear from usages whether it refers to the syntactic notion of type alias / associated type or an actual (type system) type. I tend to think about associated types as associated type aliases too.

> this PR only changes the impl-associated item, but not the trait-associated one.

That was an oversight :+1: The intention was that they were all made consistent.

**Pull request merged**